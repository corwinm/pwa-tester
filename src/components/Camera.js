import React, { useState, useEffect } from "react";
import Page from "components/Page";

function useCamera() {
  const [mediaStream, setMediaStream] = useState(undefined);

  useEffect(() => {
    if (!navigator.mediaDevices) {
      return Error("Media devices not supported.");
    }
    let videoStream;
    async function getCameraSource() {
      try {
        videoStream = await navigator.mediaDevices.getUserMedia({
          video: {
            width: { ideal: 1280 },
            height: { ideal: 720 }
          },
          audio: false
        });

        setMediaStream(videoStream);
      } catch (error) {
        console.error(error);
        setMediaStream(error);
      }
    }
    getCameraSource();

    return () => {
      if (videoStream) {
        videoStream.getVideoTracks().forEach(track => {
          track.stop();
        });
      }
    };
  }, []);

  return mediaStream;
}

export default function Camera() {
  const cameraSource = useCamera();

  if (cameraSource instanceof Error || (cameraSource && cameraSource.name)) {
    return (
      <Page title="Camera is NOT suported!">
        {cameraSource.name && cameraSource.name}
        {cameraSource.message && cameraSource.message}
      </Page>
    );
  }

  const setupCameraSource = el => {
    if (!cameraSource || !el) return;
    if ("srcObject" in el) {
      el.srcObject = cameraSource;
    } else {
      // For older browsers without the srcObject.
      el.src = window.URL.createObjectURL(cameraSource);
    }
  };

  return (
    <Page title="Camera is suported!">
      <div>Video Below?</div>
      <video id="video-chat" ref={setupCameraSource} autoPlay />
    </Page>
  );
}
