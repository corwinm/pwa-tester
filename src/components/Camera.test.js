import React from "react";
import { render, cleanup } from "react-testing-library";
import Camera from "./Camera";

jest.mock("custom-hooks/useCamera");

afterEach(cleanup);

test("matches snapshot", () => {
  const { container } = render(<Camera />);
  expect(container.firstChild).toMatchSnapshot();
});

test("Error from cameraSource shows not supported", () => {
  const { useCamera } = require("custom-hooks/useCamera");
  const testMessage = "This is a test error message";
  useCamera.mockImplementation(() => Error(testMessage));

  const { getByText, container } = render(<Camera />);

  expect(container.firstChild).toMatchSnapshot();
  expect(getByText(/Camera is NOT supported/)).toBeTruthy();
  expect(getByText(new RegExp(testMessage))).toBeTruthy();
});

test("Object from cameraSource shows supported and video", () => {
  const { useCamera } = require("custom-hooks/useCamera");
  useCamera.mockImplementation(() => {});

  const { getByText, container } = render(<Camera />);

  expect(container.firstChild).toMatchSnapshot();
  expect(getByText(/Camera is supported!/)).toBeTruthy();
  expect(container.querySelector("video")).toBeTruthy();
});
