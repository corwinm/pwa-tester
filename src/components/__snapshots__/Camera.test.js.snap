// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error from cameraSource shows not supported 1`] = `
<div
  class="notch-inset-item"
>
  <section
    class="section"
  >
    <div
      class="container is-fluid"
    >
      <h1
        class="title has-text-white"
      >
        Camera is NOT supported!
      </h1>
      <div>
        Error
        :
         
        This is a test error message
      </div>
    </div>
  </section>
</div>
`;

exports[`Object from cameraSource shows supported and video 1`] = `
<div
  class="notch-inset-item"
>
  <section
    class="section"
  >
    <div
      class="container is-fluid"
    >
      <h1
        class="title has-text-white"
      >
        Camera is supported!
      </h1>
      <div>
        <div>
          Video Below?
        </div>
        <video
          autoplay=""
          id="video-chat"
          playsinline=""
        />
      </div>
    </div>
  </section>
</div>
`;

exports[`Object from cameraSource shows supported and video with fallback video source 1`] = `
<div
  class="notch-inset-item"
>
  <section
    class="section"
  >
    <div
      class="container is-fluid"
    >
      <h1
        class="title has-text-white"
      >
        Camera is supported!
      </h1>
      <div>
        <div>
          Video Below?
        </div>
        <video
          autoplay=""
          id="video-chat"
          playsinline=""
        />
      </div>
    </div>
  </section>
</div>
`;

exports[`matches snapshot 1`] = `
<div
  class="notch-inset-item"
>
  <section
    class="section"
  >
    <div
      class="container is-fluid"
    >
      <h1
        class="title has-text-white"
      >
        Camera is supported!
      </h1>
      <div>
        <div>
          Video Below?
        </div>
        <video
          autoplay=""
          id="video-chat"
          playsinline=""
        />
      </div>
    </div>
  </section>
</div>
`;
